From 2ca2decd51ddb3653be628171c69a436ae792c92 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:48:41 -0400
Subject: [tools] perf/record: Default -t option to no inheritance

Message-id: <1408953210-26343-34-git-send-email-jolsa@redhat.com>
Patchwork-id: 88780
O-Subject: [PATCH RHEL7.1 BZ1133083 033/322] perf record: Default -t option to no inheritance
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit 69e7e5b02bc6a9e5cf4a54911b27ca133cc1f99f
Author: Adrian Hunter <adrian.hunter@intel.com>
Date: Mon Nov 18 11:55:57 2013 +0200

description
===========
The change to per-cpu mmaps causes the -p, -t and -u options now to have
inheritance enabled by default.  Change that back to no inheritance but
for the -t option only.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/Documentation/perf-record.txt b/tools/perf/Documentation/perf-record.txt
index 6ac867e..c407897 100644
--- a/tools/perf/Documentation/perf-record.txt
+++ b/tools/perf/Documentation/perf-record.txt
@@ -57,6 +57,8 @@ OPTIONS
 -t::
 --tid=::
         Record events on existing thread ID (comma separated list).
+        This option also disables inheritance by default.  Enable it by adding
+        --inherit.
 
 -u::
 --uid=::
diff --git a/tools/perf/builtin-record.c b/tools/perf/builtin-record.c
index f5b18b8..65615a8 100644
--- a/tools/perf/builtin-record.c
+++ b/tools/perf/builtin-record.c
@@ -843,8 +843,9 @@ const struct option record_options[] = {
  OPT_U64('c', "count", &record.opts.user_interval, "event period to sample"),
  OPT_STRING('o', "output", &record.file.path, "file",
       "output file name"),
- OPT_BOOLEAN('i', "no-inherit", &record.opts.no_inherit,
-      "child tasks do not inherit counters"),
+ OPT_BOOLEAN_SET('i', "no-inherit", &record.opts.no_inherit,
+   &record.opts.no_inherit_set,
+   "child tasks do not inherit counters"),
  OPT_UINTEGER('F', "freq", &record.opts.user_freq, "profile at this frequency"),
  OPT_CALLBACK('m', "mmap-pages", &record.opts.mmap_pages, "pages",
        "number of mmap data pages",
@@ -939,6 +940,9 @@ int cmd_record(int argc, const char **argv, const char *prefix __maybe_unused)
   goto out_symbol_exit;
  }
 
+ if (rec->opts.target.tid && !rec->opts.no_inherit_set)
+  rec->opts.no_inherit = true;
+
  err = target__validate(&rec->opts.target);
  if (err) {
   target__strerror(&rec->opts.target, err, errbuf, BUFSIZ);
diff --git a/tools/perf/perf.h b/tools/perf/perf.h
index b079304..b23fed5 100644
--- a/tools/perf/perf.h
+++ b/tools/perf/perf.h
@@ -254,6 +254,7 @@ struct perf_record_opts {
  bool      inherit_stat;
  bool      no_delay;
  bool      no_inherit;
+ bool      no_inherit_set;
  bool      no_samples;
  bool      raw_samples;
  bool      sample_address;
-- 
1.7.1